x/gov/keeper.go

```diff
@@ -0,0 +1,415 @@
+	if err != nil {
+		panic(err)
+	}
+
+	key := []byte(fmt.Sprintf("%d", proposal.ProposalID) + ":proposal")
+
+	store.Set(key, bz)
+}
+
+// Implements sdk.AccountMapper.
+func (keeper Keeper) DeleteProposal(ctx sdk.Context, proposal *Proposal) {
+	store := ctx.KVStore(keeper.storeKey)
+	key := []byte(fmt.Sprintf("%d", proposal.ProposalID) + ":proposal")
+	store.Delete(key)
+}
+
+func (keeper Keeper) getNewProposalID(ctx sdk.Context) (proposalID int64) {
+	store := ctx.KVStore(keeper.storeKey)
+	bz := store.Get([]byte("newProposalID"))
+	if bz == nil {
+		proposalID = 1
+	} else {
+		keeper.cdc.MustUnmarshalBinary(bz, &proposalID) // TODO: switch to UnmarshalBinaryBare when new go-amino gets added
+	}
+	bz = keeper.cdc.MustMarshalBinary(proposalID + 1) // TODO: switch to MarshalBinaryBare when new go-amino gets added
+	store.Set([]byte("newProposalID"), bz)
+	return proposalID
+}
+
+func (keeper Keeper) activateVotingPeriod(ctx sdk.Context, proposal *Proposal) {
+	proposal.VotingStartBlock = ctx.BlockHeight()
+	proposal.Status = "VotingPeriod"
+	keeper.SetProposal(ctx, proposal)
+	keeper.ActiveProposalQueuePush(ctx, proposal)
+}
+
+// =====================================================
+// Procedures
+
+// Gets procedure from store. TODO: move to global param store and allow for updating of this
+func (keeper Keeper) GetDepositProcedure(ctx sdk.Context) *DepositProcedure {
+	return &DepositProcedure{
+		MinDeposit:       sdk.Coins{{"steak", 10}},
+		MaxDepositPeriod: 200,
+	}
+}
+
+// Gets procedure from store. TODO: move to global param store and allow for updating of this
+func (keeper Keeper) GetVotingProcedure(ctx sdk.Context) *VotingProcedure {
+	return &VotingProcedure{
+		VotingPeriod: 200,
+	}
+}
+
+// Gets procedure from store. TODO: move to global param store and allow for updating of this
+func (keeper Keeper) GetTallyingProcedure(ctx sdk.Context) *TallyingProcedure {
+	return &TallyingProcedure{
+		Threshold:         sdk.NewRat(1, 2),
+		Veto:              sdk.NewRat(1, 3),
+		GovernancePenalty: sdk.NewRat(1, 100),
+	}
+}
+
+// =====================================================
+// Votes
+
+// Gets the vote of a specific voter on a specific proposal
+func (keeper Keeper) AddVote(ctx sdk.Context, proposalID int64, voter sdk.Address, option string) sdk.Error {
+	proposal := keeper.GetProposal(ctx, proposalID)
+	if proposal == nil {
+		return ErrUnknownProposal(proposalID)
+	}
+	if proposal.Status != "VotingPeriod" {
+		return ErrInactiveProposal(proposalID)
+	}
+
+	if option != "Yes" && option != "Abstain" && option != "No" && option != "NoWithVeto" {
+		return ErrInvalidVote(option)
+	}
+
+	vote := Vote{
+		ProposalID: proposalID,
+		Voter:      voter,
+		Option:     option,
+	}
+	keeper.setVote(ctx, proposalID, voter, vote)
+
+	return nil
+}
+
+// Gets the vote of a specific voter on a specific proposal
+func (keeper Keeper) GetVote(ctx sdk.Context, proposalID int64, voter sdk.Address) *Vote {
```

Comment:
Any particular reason these functions return pointers instead of structs?

Reply:
My biggest comment is that these need to be turned into non-pointer objects